src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs

```diff
@@ -203,15 +212,17 @@ internal JsonPropertyInfo(JsonPropertyInfoValues<T> propertyInfo, JsonSerializer
 
         private protected override void DetermineEffectiveConverter(JsonTypeInfo jsonTypeInfo)
         {
-            JsonConverter converter =
-                Options.ExpandConverterFactory(CustomConverter, PropertyType)
-                ?? jsonTypeInfo.Converter;
+            Debug.Assert(jsonTypeInfo is JsonTypeInfo<T>);
```

Comment:
What guarantees this?

Reply:
See https://github.com/dotnet/runtime/pull/73259/files/66b37e4903a264bd3d63b39640e09f3c2be07d51#r937637507. The callsite is always guaranteed to pass a JsonTypeInfo for the property type (which is now guaranteed to match `T`):

https://github.com/dotnet/runtime/blob/8daeec27929d15576f034a8743c2c9c524d24609/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs#L295-L296